src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2172,10 +2172,944 @@ - (void)doTestSubscriptionPoolWithSize:(NSInteger)subscriptionPoolSize
+        dispatch_sync(self->_storageQueue, ^{
+            XCTAssertTrue([[controller.controllerDataStore _fetchEndpointIndexForNodeID:deviceID] isEqualToArray:testEndpoints]);
+            for (NSNumber * endpoint in testEndpoints) {
+                [initialClusterIndex insertObject:[controller.controllerDataStore _fetchClusterIndexForNodeID:deviceID endpointID:endpoint] atIndex:endpoint.unsignedShortValue];
+                XCTAssertNotNil(initialClusterIndex);
+                for (NSNumber * cluster in [initialClusterIndex objectAtIndex:endpoint.unsignedShortValue]) {
+                    XCTAssertNotNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                    MTRClusterPath * path = [MTRClusterPath clusterPathWithEndpointID:endpoint clusterID:cluster];
+                    XCTAssertTrue([device _persistedClusterContains:path]);
+                    XCTAssertNotNil([device _getPersistedClusterDataForPath:path]);
+                }
+            }
+        });
+        [subscriptionExpectation fulfill];
+    };
+
+    [device setDelegate:delegate queue:queue];
+
+    [self waitForExpectations:@[ subscriptionExpectation ] timeout:60];
+
+    // Inject a fake attribute report adding a cluster to endpoint 1 to the server list.
+    dataVersionForServerList = [NSNumber numberWithUnsignedLongLong:(dataVersionForServerList.unsignedLongLongValue + 1)];
+    [testClusterDataValue addObject:@{
+        MTRDataKey : @ {
+            MTRTypeKey : MTRUnsignedIntegerValueType,
+            MTRValueKey : @(5),
+        }
+    }];
+
+    NSArray<NSDictionary<NSString *, id> *> * attributeReport = @[ @{
+        MTRAttributePathKey : [MTRAttributePath attributePathWithEndpointID:testEndpoint clusterID:@(MTRClusterIDTypeDescriptorID) attributeID:@(MTRAttributeIDTypeClusterDescriptorAttributeServerListID)],
+        MTRDataKey : @ {
+            MTRDataVersionKey : dataVersionForServerList,
+            MTRTypeKey : MTRArrayValueType,
+            MTRValueKey : testClusterDataValue,
+        }
+    } ];
+
+    XCTestExpectation * attributeDataReceivedExpectation = [self expectationWithDescription:@"Injected Attribute data received"];
+    XCTestExpectation * reportEndExpectation = [self expectationWithDescription:@"Injected Attribute data report ended"];
+    delegate.onAttributeDataReceived = ^(NSArray<NSDictionary<NSString *, id> *> * attributeReport) {
+        attributeReportsReceived += attributeReport.count;
+        XCTAssert(attributeReportsReceived > 0);
+        [attributeDataReceivedExpectation fulfill];
+    };
+
+    delegate.onReportEnd = ^{
+        // Verify that the data store is populated with cluster index and cluster data for endpoints 0, 1 and 2.
+        // Also _persistedClusters and _persistedClusterData has the cluster paths for all clusters in the cluster index. Nothing is deleted.
```

Comment:
But we're not checking the new thing got added?

It seems a bit odd to me the way these tests are structured: we have a test where we add a thing and check that nothing got removed, then a very similar test where we remove one thing and check that the thing got removed.

It seems to me like we should have a "here is the set of changes to make, then test that they took effect" helper with one copy of all this verification code, then we call it a few times with various scenarios: add a thing, remove a thing, add a thing while removing a different thing.  It would be a lot less code and much clearer, and have more test coverage and make it easy to add even more...

Reply:
filed #33645 to track for add a thing. fixed for remove a thing